### PR TITLE
add build option for adiak

### PR DESCRIPTION
**Pull request by slabasan**
FYI: Update adiak call to `adiak::collect_all()` once adiak is updated to 0.4.0.

Metadata before:
```
$ CALI_CONFIG=runtime-report,print.metadata ./ams_example --precision double --uqtype faiss-mean -S ../../tests/AMSlib/debug_model.pt -H ../../tests/AMSlib/example_faiss.idx -e 100

mpi.world.size       : 1 
cali.caliper.version : 2.9.0
cali.channel         : runtime-report
...
```

Metadata after:
```
$ CALI_CONFIG=runtime-report,print.metadata ./ams_example --precision double --uqtype faiss-mean -S ../../tests/AMSlib/debug_model.pt -H ../../tests/AMSlib/example_faiss.idx -e 100

mpi.world.size       :          1 
cali.caliper.version : 2.9.0
walltime             :   5.814216 
systime              :   1.040000 
cputime              :  33.830000 
compiler             : @RAJAPERF_COMPILER@
numhosts             :          1 
cluster              : lassen
hostname             : lassen23
cmdline              : [./ams_example,--precisio~~example_faiss.idx,-e,100]
libraries            : [.../lib64/libnss_files.so.2]
working_directory    : <redacting>
executablepath       : <redacting>
executable           : ams_example
launchday            : 1709683200 
launchdate           : 1709757169 
uid                  : Stephanie
cali.channel         : runtime-report
...
```

### CODE REVIEW

**Comment by slabasan** [comment]
Supercedes #10 

@koparasy This is ready for your review

File: examples/main.cpp [hunk 1]
```diff
@@ -642,6 +643,28 @@ int main(int argc, char **argv)
...
+  adiak::hostname();
+  adiak::clustername();
+  adiak::walltime();
+  adiak::systime();
+  adiak::cputime();
+  adiak::jobsize();
+  adiak::hostlist();
+  adiak::numhosts();
+  adiak::value("compiler", std::string("@RAJAPERF_COMPILER@"));
+
```
Comment (koparasy) [inline examples/main.cpp]: Shouldn't these function calls be guarded with` #ifdef __ENABLE_ADIAK__` and `#endif` respectively?

File: examples/main.cpp [hunk 2]
```diff
@@ -876,6 +899,10 @@ int main(int argc, char **argv)
     return -1;
   }
 
+  // ---------------------------------------------------------------------------
+  // adiak finalize
+  adiak::fini();
```
Comment (koparasy) [inline examples/main.cpp:909]: Guard with `__ENABLE_ADIAK__`

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -52,6 +52,7 @@ jobs:
             -DFAISS_DIR=$AMS_FAISS_PATH \
             -DWITH_AMS_DEBUG=On \
             -DWITH_WORKFLOW=On \
+            -DWITH_ADIAK=On \
```
Comment (koparasy) [inline .github/workflows/ci.yml:55]: Can you please add a test case in which ADIAK is off?

**Review by koparasy — CHANGES_REQUESTED** [review]
This overall looks good to me. Thank you for updating the tests as well. Could you please fix the minor comments? I will merge after.

File: examples/main.cpp (hunk 1 above)
Comment (slabasan) [reply examples/main.cpp]: Yes, good catch! Added the guards.

File: .github/workflows/ci.yml (hunk 3 above)
Comment (slabasan) [reply .github/workflows/ci.yml:55]: Done, updated one config in each job for adiak to be off

**Review by koparasy — APPROVED** [review]
This looks great! Thank you for setting adiak by default off.

**Pull request merged**